iceoryx_posh/source/roudi/service_registry.cpp

```diff
@@ -16,45 +17,163 @@
-    cxx::set::remove(m_serviceMap[service].instanceSet, instance);
+    bool removedElement{false};
+
+    uint64_t index = 0U;
+    for (auto iterator = m_serviceDescriptionVector.begin(); iterator != m_serviceDescriptionVector.end();)
+    {
+        if (m_serviceDescriptionVector[index] == serviceDescription)
+        {
+            uint64_t removedValue{0U};
+            bool removedEntry{false};
+            for (auto it = m_serviceMap.begin(); it != m_serviceMap.end();)
+            {
+                if (it->second == index)
+                {
+                    removedValue = it->second;
+                    it = m_serviceMap.erase(it);
+                    removedEntry = true;
+                    continue;
+                }
+                else if (removedEntry && it->second > removedValue)
+                {
+                    // update index due to removed element
+                    it->second--;
+                }
+                it++;
+            }
+
+            removedValue = 0U;
+            removedEntry = false;
+            for (auto it = m_instanceMap.begin(); it != m_instanceMap.end();)
+            {
+                if (it->second == index)
+                {
+                    removedValue = it->second;
+                    it = m_instanceMap.erase(it);
+                    removedEntry = true;
+                    continue;
+                }
+                else if (removedEntry && it->second > removedValue)
+                {
+                    // update index due to removed element
+                    it->second--;
+                }
+                it++;
+            }
+            m_serviceDescriptionVector.erase(iterator);
+            removedElement = true;
+            // There can be not more than one element
+            break;
+        }
+        index++;
+        iterator++;
+    }
+
+    return removedElement;
 }
 
-void ServiceRegistry::find(InstanceSet_t& instances,
+void ServiceRegistry::find(ServiceDescriptionVector_t& searchResult,
                            const CaproIdString_t& service,
-                           const CaproIdString_t& instance) const
+                           const CaproIdString_t& instance) const noexcept
 {
-    if (instance == iox::cxx::string<100>(capro::AnyInstanceString))
+    cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> intersection;
+
+    // Find (K1, K2)
+    // O(log n + log n + max(#PossibleServices + #possiblesInstances) + #intersection)
+    if (instance != capro::IdString_t(capro::AnyInstanceString)
+        && service != capro::IdString_t(capro::AnyServiceString))
     {
-        for (auto& instance : m_serviceMap[service].instanceSet)
+        cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> possibleServices;
+        cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> possibleInstances;
+
+        auto rangeServiceMap = m_serviceMap.equal_range(service);
+        for (auto entry = rangeServiceMap.first; entry != rangeServiceMap.second; ++entry)
```

Comment:
cheers for inconsistencies in the STL for not being able to use range-based for loops here